sections/data_release_production_pipelines.tex

```diff
@@ -71,26 +75,332 @@ \section{Data Release Production}
+
+The final output data products of the ImChar/JointCal iteration are the Source table and CalExp (calibrated exposure) images.  The latter includes Image, Mask, Variance, Background, PSF, WCS, Calib components that we will track separately.
+
+There are also several intermediate versions of the Source and CalExp data products passed between the ImChar/JointCal pipelines, as well as...
+
+\begin{note}[TODO]
+  finish discussing data products.
+\end{note}
+
 \subsubsection{BootstrapImChar}
 \label{sec:drpBootstrapImChar}
-\subsubsection{BootstrapJointCal}
-\label{sec:drpBootstrapJointCal}
+
+The BootstrapImChar pipeline is the first thing run on each science exposure in a data release.  It has the difficult task of bootstrapping multiple quantities (PSF, WCS, photometric calibration, background model, etc.) that each normally require all of the others to be specified when one is fit.  As a result, while the algorithmic components to be run in this pipeline are generally clear, their ordering and specific requirements are not; algorithms that are run early will have a harder task than algorithms that are run later, and some iteration will almost certainly be necessary.
+
+A plausible (but by no means certain) high-level algorithm for this pipeline is given below in pseudocode.  Highlighted terms are described in more detail below the pseudocode block.
+
+\lstset{
+    language=Python,
+    basicstyle=\scriptsize\ttfamily,
+    keywordstyle=\bfseries,
+    commentstyle=\color{darkgray},
+    escapeinside={\%}{\%},
+}
+
+% Define a local macro that lets us refer to sections of the text
+% more easily (will undefine at the end of this section).
+\newcommand{\hr}[1]{\hyperref[sec:drpBootstrapImChar_#1]{#1}}
+
+\begin{lstlisting}
+def BootstrapImChar(%\hr{raw}%, %\hr{reference}%):
+    # Some data products components are visit-wide and some are per-CCD;
+    # these imaginary data types lets us deal with both.
+    # VisitExposure also has components; most are self-explanatory, and
+    # {mi} == {image,mask,variance} (for "MaskedImage").
+    calexp = VisitExposure()
+    sources = VisitCatalog()
+    snaps = VisitMaskedImageList()  # holds both snaps, but only {image,mask,variance}
+    parallel for ccd in ALL_SENSORS:
+        snaps[ccd] = [%\hr{RunISR}%(raw[ccd]) for snap in SNAP_NUMBERS]
+        snaps[ccd].mask = %\hr{SubtractSnaps}%(snaps[ccd])
+        calexp[ccd].mi = %\hr{CombineSnaps}%(snaps[ccd])
+    calexp.psf = %\hr{FitWavefront}%(calexp[WAVEFRONT_SENSORS].mi)
+    calexp.{image,mask,variance,background}
+        = %\hr{SubtractBackground}%(calexp.mi)
+    parallel for ccd in ALL_SENSORS:
+        sources[ccd] = %\hr{DetectSources}%(calexp.{mi,psf})
+    sources[ccd] = %\hr{DeblendSources}%(sources[ccd], calexp.{mi,psf})
+    sources[ccd] = %\hr{MeasureSources}%(sources[ccd], calexp.{mi,psf})
+    matches = %\hr{MatchSemiBlind}%(sources, reference)
+    while not converged:
```

Comment:
Define "converged". Is it guaranteed within $SMALLNUM iterations? Can we set an upper limit on the number of iterations required, and if we hit it either drop the data or press on regardless?